### PR TITLE
Add a library for backoff logic.

### PR DESCRIPTION
**Pull request by vishnu-deepsource**
The backoff library currently implements two methods:
1. Exponential backoff.
2. Fibonacci backoff.

### CODE REVIEW

File: backoff/fibonacci.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	if b.mem == nil {
+		b.mem = &map[uint64]uint64{}
+	}
+	m := *b.mem
+	if val, ok := m[n]; ok {
+		return val
+	}
+	if n <= 1 {
+		return n
+	}
```
Comment (sourya-deepsource) [inline backoff/fibonacci.go:81]: Handle the `n=0` and `n=1` cases separately.

**Review by sourya-deepsource — CHANGES_REQUESTED** [review]

File: backoff/exponential.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sync/atomic"
+	"time"
+)
+
+type Exponential struct {
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:11]: ExponentialBackoff would stutter since the type would become backoff.ExponentialBackoff.

File: backoff/exponential.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+func (b *Exponential) Duration() time.Duration {
+	d := b.ForAttempt(float64(atomic.AddUint64(&b.attempt, 1) - 1))
+	return d
+}
+
+const maxInt64 = float64(math.MaxInt64 - 512)
+
+func (b *Exponential) ForAttempt(attempt float64) time.Duration {
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:28]: Agreed.  Will fix.

File: backoff/exponential.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sourya-deepsource) [reply backoff/exponential.go:11]: Makes sense.

File: backoff/exponential.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	Factor float64
+
+	Jitter bool
+
+	Min, Max time.Duration
+}
+
+func (b *Exponential) Duration() time.Duration {
+	d := b.ForAttempt(float64(atomic.AddUint64(&b.attempt, 1) - 1))
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:22]: Probably not, since this is the function that modifies b.attempt.  The ForAttempt can be used externally too.

File: backoff/exponential.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	return d
+}
+
+const maxInt64 = float64(math.MaxInt64 - 512)
+
+func (b *Exponential) ForAttempt(attempt float64) time.Duration {
+
+	min := b.Min
+	fmt.Println(min)
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:31]: Oops.

File: backoff/exponential.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	if min >= max {
+		return max
+	}
+
+	factor := b.Factor
+	if factor <= 0 {
+		factor = 2
+	}
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:50]: Yup.  Agreed.

File: backoff/exponential.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+	minf := float64(min)
+	durf := minf * math.Pow(factor, attempt)
+
+	if b.Jitter {
+		durf = rand.Float64()*(durf-minf) + minf
+	}
+
+	if durf > maxInt64 {
+		return max
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:60]: Agreed.

File: backoff/exponential.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	dur := time.Duration(durf)
+
+	if dur > max {
+		return max
+	}
+	return dur
+}
+
+func (b *Exponential) Reset() {
+	atomic.StoreUint64(&b.attempt, 0)
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential.go:72]: Trying to make this thread-safe.

File: backoff/fibonacci.go (hunk 1 above)
Comment (vishnu-deepsource) [reply backoff/fibonacci.go:81]: Didn't see the point..

File: backoff/exponential.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sourya-deepsource) [reply backoff/exponential.go:22]: I mean, the conversion to float64.

File: backoff/fibonacci.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	"math/rand"
+	"sync/atomic"
+	"time"
+)
+
+type Fibonacci struct {
+	attempt  uint64
+	Jitter   bool
+	Min, Max time.Duration
+	mem      *map[uint64]uint64
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/fibonacci.go:13]: Agreed.

File: backoff/fibonacci_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+	tests := []struct {
+		name   string
+		fields fields
+		want   time.Duration
+	}{
+		{
+			"must generate the correct duration for attempt=6",
+			fields{
+				attempt: uint64(6),
+				Jitter:  false,
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/fibonacci_test.go:24]: Yup.  Good catch.

File: backoff/exponential_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+		name   string
+		fields fields
+		want   time.Duration
+	}{
+		{
+			"must generate the correct duration for attempt=3",
+			fields{
+				attempt: uint64(2),
+				Factor:  float64(3),
+				Jitter:  false,
```
[the remark this answers is not in the document]
Comment (vishnu-deepsource) [reply backoff/exponential_test.go:26]: Good catch.

**Comment by souvik-deepsource** [comment]
@vishnu-deepsource Probably add tests when `Jitter` is enabled. The duration returned should be within the acceptable range.

**Comment by vishnu-deepsource** [comment]
@souvik-deepsource Addressing them.

File: backoff/fibonacci_test.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+		Jitter  bool
+		Min     time.Duration
+		Max     time.Duration
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		want   time.Duration
+	}{
+		{
```
Comment (burntcarrot) [inline backoff/fibonacci_test.go:20]: Apologies for hijacking the review; one suggestion: we could add a test for `attempt=0`, just to make sure that the returned value is 0 and not any other non-zero value.

```
{
	"must generate the correct duration for attempt=0",
	fields{
		attempt: uint64(0),
		Min:     1 * time.Minute,
		Max:     10 * time.Minute,
	},
	0,
},
```

Could add a similar test for `attempt=1`, in order to cover [this](https://github.com/deepsourcelabs/hermes/pull/3/files#diff-6ceb2d7871c325d41e01873bcc312fbb85dea5bb686b29b2ad9b6b33e0c63166R79) piece of return logic.

Comment (vishnu-deepsource) [reply backoff/fibonacci_test.go:50]: One question is if attempt=0, shouldn't we return the Min instead?

Comment (vishnu-deepsource) [reply backoff/fibonacci_test.go:50]: Agreed.  We do send 0 currently.  Resolving this for now.

**Review by sourya-deepsource — APPROVED** [review]
